FeedStoreChallenge/FeedStore/CoreData/CoreDataFeedStore.swift

```
@@ -0,0 +1,125 @@
+        }
+        
+        try loadError.map {
+            throw LoadingError.failedToLoadPersistentStores($0)
+        }
+        
+        managedContext = storeContainer.newBackgroundContext()
+    }
+    
+    public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+        do {
+            if let currentCache = try CDCache.fetchCachedFeed(managedContext) {
+                managedContext.delete(currentCache)
+                
+                if let saveError = saveContext() {
+                    completion(.some(saveError))
+                } else {
+                    completion(.none)
+                }
+            } else {
+                completion(.none)
+            }
+        } catch {
+            completion(.some(error))
+        }
+    }
+    
+    public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+        do {
+            if let currentCache = try CDCache.fetchCachedFeed(managedContext) {
+                managedContext.delete(currentCache)
+            }
+        } catch { /* If there is an error then just insert the feed in Core Data*/ }
+        
+        let cdFeed = mapLocalFeedToCoreDataFeed(feed, timestamp: timestamp)
+        
+        if let saveError = saveContext() {
+            completion(.some(saveError))
+        } else {
+            completion(.none)
+        }
+    }
+    
+    public func retrieve(completion: @escaping RetrievalCompletion) {
+        do {
+            if let cache = try CDCache.fetchCachedFeed(managedContext) {
+                let imageFeed = cache.feed.compactMap { ($0 as? CDFeedImage)?.local }
+                completion(.found(feed: imageFeed, timestamp: cache.timestamp))
+            } else {
+                completion(.empty)
+            }
+        } catch {
+            completion(.failure(error))
+        }
+    }
+    
+    
+    
+    // MARK: Helper Methods
+    
+    private func saveContext() -> Error? {
```

This method can `throw` instead of returning an optional `Error?`. (throwing errors is more idiomatic in Swift)